python/paddle/distributed/auto_parallel/api.py

```diff
@@ -733,6 +769,139 @@ def __getattr__(self, item):
         return getattr(self._inner_opt, item)
 
 
+class ShardingStage1Fn:
+    """
+    ShardOptimizer shard_fn for stage 1, user can use this to implement sharding stage 1
+    Args:
+        key (str): The name of optimizer state. For example, it can be beta, momentum, etc.
+        param (paddle.Tensor): The parameter which correspond to the input accumulator
+        accumulator (paddle.Tensor): The optimizer state need to be shard of this parameter
+
+    Returns:
+        None
+
+    Examples:
+        .. code-block:: python
+            >>> import paddle
+            >>> import paddle.distributed as dist
+            >>> mesh = dist.ProcessMesh([0, 1], dim_names=["x"])
+            >>> class MLP(paddle.nn.Layer):
+            ...     def __init__(self):
+            ...         super().__init__()
+            ...         self.fc1 = paddle.nn.Linear(8, 8)
+            ...         self.fc2 = paddle.nn.Linear(8, 8)
+            ...
+            ...     def forward(self, input):
+            ...         return self.fc2(self.fc1(input))
+
+            >>> # doctest: +REQUIRES(env:DISTRIBUTED)
+            >>> layer = MLP()
+            >>> batch = paddle.rand(shape=[8, 8])
+            >>> opt = paddle.optimizer.AdamW(parameters=layer.parameters())
+            >>> opt = dist.shard_optimizer(opt, dist.ShardingStage1Fn())
+            >>> for _ in range(5):
+            >>>     loss = layer(batch)
+            >>>     loss.backward()
+            >>>     opt.step()
+            >>>     opt.clear_grad()
+            >>> # This case need to be executed in multi-card environment
+            >>> # python -m paddle.distributed.launch --gpus=0,1 {test_case}.py
+    """
+
+    def __call__(self, key, param, accumulator):
+        if param.is_dist():
+            # Only deal with momentum in optimizer, beta should be replicated cross param's mesh
+            if 'beta' not in key:
+                placements = get_placement_with_sharding(param)
+            else:
+                placements = [
+                    dist.Replicate()
+                    for _ in range(len(param.process_mesh.shape))
+                ]
+            return shard_tensor(
+                accumulator,
+                mesh=param.process_mesh,
+                placements=placements,
+            )
+        return accumulator
+
+
+class ShardingStage3Fn:
+    """
+    ShardOptimizer shard_fn for stage 3, user can use this to implement sharding stage 3
```

Comment:
A builtin shard_fn for ``shard_optimizer`` interface, users can pass it to ``shard_optimizer`` to implement sharding optimization with stage 3.

Reply:
done